### PR TITLE
util/cmpver: faster Compare (-85% runtime)

### PR DESCRIPTION
**Pull request by icio**
This comes at the cost of being much harder to read and reason about. I'm open to opinions on whether this is a worthwhile trade-off.

Benchmark comparing before/after eb773f99:

```
goos: darwin
goarch: arm64
pkg: tailscale.com/util/cmpver
                                                    │ base@2e861c35.log │ build.icio_cmpver-faster@eb773f99.log │
                                                    │      sec/op       │     sec/op      vs base               │
Compare/both_empty-12                                       1.991n ± 2%      1.982n ± 3%        ~ (p=0.900 n=6)
Compare/v1_empty-12                                        18.385n ± 4%      1.947n ± 2%  -89.41% (p=0.002 n=6)
Compare/v2_empty-12                                        18.430n ± 7%      1.932n ± 3%  -89.51% (p=0.002 n=6)
Compare/semver_major-12                                    26.390n ± 4%      5.435n ± 3%  -79.41% (p=0.002 n=6)
Compare/semver_major#01-12                                 26.250n ± 3%      5.420n ± 3%  -79.35% (p=0.002 n=6)
Compare/semver_minor-12                                     61.44n ± 4%      10.29n ± 4%  -83.25% (p=0.002 n=6)
Compare/semver_minor_missing-12                            51.565n ± 4%      5.215n ± 3%  -89.89% (p=0.002 n=6)
Compare/semver_patch-12                                     93.48n ± 5%      13.25n ± 5%  -85.83% (p=0.002 n=6)
Compare/semver_patch_missing-12                            86.840n ± 5%      9.211n ± 3%  -89.39% (p=0.002 n=6)
Compare/semver_equal-12                                     94.07n ± 5%      12.47n ± 5%  -86.74% (p=0.002 n=6)
Compare/tailscale_major-12                                 26.230n ± 4%      4.066n ± 3%  -84.50% (p=0.002 n=6)
Compare/tailscale_minor-12                                  67.36n ± 4%      11.85n ± 4%  -82.41% (p=0.002 n=6)
Compare/tailscale_patch-12                                 111.95n ± 3%      17.18n ± 4%  -84.65% (p=0.002 n=6)
Compare/tailscale_equal-12                                 111.95n ± 4%      16.10n ± 6%  -85.62% (p=0.002 n=6)
Compare/tailscale_weird_extra_field-12                      88.45n ± 4%      10.68n ± 4%  -87.93% (p=0.002 n=6)
Compare/only_ascii_numbers-12                              15.455n ± 5%      2.942n ± 2%  -80.96% (p=0.002 n=6)
Compare/zeroes_equal-12                                    33.845n ± 4%      5.390n ± 4%  -84.07% (p=0.002 n=6)
Compare/zeroes_nested_equal-12                              74.69n ± 4%      12.05n ± 5%  -83.86% (p=0.002 n=6)
Compare/windows_version-12                                 120.60n ± 4%      19.33n ± 4%  -83.98% (p=0.002 n=6)
Compare/windows_11_is_everything_above_10.0.22000-12       120.40n ± 3%      19.06n ± 4%  -84.17% (p=0.002 n=6)
Compare/android_short_version-12                           25.745n ± 5%      4.617n ± 3%  -82.06% (p=0.002 n=6)
Compare/android_longer_version-12                          34.985n ± 4%      4.330n ± 3%  -87.62% (p=0.002 n=6)
Compare/iOS_version-12                                     77.420n ± 3%      9.532n ± 4%  -87.69% (p=0.002 n=6)
Compare/Linux_short_kernel_version-12                      60.645n ± 3%      7.319n ± 4%  -87.93% (p=0.002 n=6)
Compare/Linux_long_kernel_version-12                       64.495n ± 3%      7.311n ± 4%  -88.67% (p=0.002 n=6)
Compare/FreeBSD_version-12                                 44.035n ± 3%      5.114n ± 4%  -88.39% (p=0.002 n=6)
Compare/Synology_version-12                                 64.77n ± 3%      13.07n ± 5%  -79.83% (p=0.002 n=6)
Compare/chars-12                                                             4.865n ± 4%
geomean                                                     46.38n           7.065n       -84.55%
```

### CODE REVIEW

File: util/cmpver/version_test.go [hunk 1]
```diff
@@ -9,156 +9,189 @@ import (
...
+	},
+	{
+		name: "Linux short kernel version",
+		v1:   "4.4.302+",
+		v2:   "4.0",
+		want: 1,
+	},
+	{
+		name: "Linux long kernel version",
+		v1:   "4.14.255-311-248.529.amzn2.x86 64",
```
Comment (bradfitz) [inline util/cmpver/version_test.go:176]: why this test change from underscore to space?

**Review by bradfitz — COMMENTED** [review]
This new code's pretty hard to follow (which would normally make me err on passing on this for simpler code) but the old code isn't a walk in the park either.

Still, let's only do this if this is profiles say it's really hot and worth it.

Also, if we do this I'd then want a fuzz test for the new code, fuzzing against the old code (which could live only in a test file) to make sure the old and new implementations match. (If we find they diverge in a fuzz failure we can then add that to the test table as well.)

File: util/cmpver/version_test.go (hunk 1 above)
Comment (icio) [reply util/cmpver/version_test.go:176]: Find and replace error - the correct usage of VSCode's "replace in selection" continues to evade me.
